### PR TITLE
Allow setting all cache entries for force caching

### PR DESCRIPTION
**Pull request by rsanheim**
**Story card:** [ch2653](https://app.clubhouse.io/simpledotorg/story/2653/make-sure-we-are-force-setting-the-cache-w-multi-entry-calls)

Turns out `fetch_multi` doesn't respect the `force` keyword. So we have to handle that ourselves. This does that branching ourselves, and uses `write_multi`, which is a `MSET` under the hood, to make sure we set all new entries. Tracking the Rails bug at https://github.com/rails/rails/issues/41535.

This *also* removes `redis-rails`, because [Redis cache support is built in to Rails as of 5.2](https://blog.engineyard.com/rails-5-2-redis-cache-store) _and_ redis-rails doesn't support MSET.  Including `redis-rails` confuses things for us, because it isn't clear what code is actually happening under the hood when we are doing Rails.cache calls.  This also will better prepare the way for Rails 6.x and beyond.  See also https://bigbinary.com/blog/rails-5.2-adds-write_multi-for-cache-writes for more context on this change.

Because I removed the `Redis-rails` gem, there is a very small change to how we configure the Redis cache servers...the config expects a URL now, not a host. I made that change in https://github.com/simpledotorg/deployment/pull/312, which will need to be deployed before this PR gets merged + shipped.

### CODE REVIEW

File: .env.development [hunk 1]
```diff
@@ -11,7 +11,7 @@ DATABASE_WITH_PHI_PASSWORD=
 DATABASE_WITH_PHI_USERNAME=
 
 RAILS_CACHE_REDIS_PASSWORD=
-RAILS_CACHE_REDIS_URL=localhost
+RAILS_CACHE_REDIS_URL="redis://localhost:6379/0"
```
Comment (rsanheim) [inline .env.development:14]: This is the dev config change needed for the updated, built in Redis caching support.  The deployment changes are the same, but just w/ the ElasticCache URLs instead.

File: app/models/reports/repository.rb [hunk 2]
```diff
@@ -45,9 +45,13 @@ def no_bp_measure_count
     #
     def cached_query(calculation, &block)
       items = cache_keys(calculation)
-      cached_results = cache.fetch_multi(*items, force: force_cache?) { |item|
-        block.call(item)
-      }
+      cached_results = if force_cache? # fetch_multi does not support force, so we do it manually here
+        cached_results = items.each_with_object({}) { |item, hsh| hsh[item] = block.call(item) }
```
Comment (prabhanshuguptagit) [inline app/models/reports/repository.rb:49]: Shall we extract this into a helper method that takes a block? We might have to do a `fetch_multi` in more places.

File: spec/services/control_rate_service_spec.rb [hunk 3]
```diff
@@ -333,7 +333,7 @@ def refresh_views
   end
 
   context "caching" do
-    let(:memory_store) { ActiveSupport::Cache.lookup_store(:redis_store) }
+    let(:memory_store) { ActiveSupport::Cache.lookup_store(:redis_cache_store) }
```
Comment (prabhanshuguptagit) [inline spec/services/control_rate_service_spec.rb:336]: 👾  I think I wrote that, this should actually be
```suggestion
    let(:memory_store) { ActiveSupport::Cache.lookup_store(:memory_store) }
```

**Review by prabhanshuguptagit — CHANGES_REQUESTED** [review]
✨  Looks solid. There's a usage of `fetch_multi` in `NoBPMeasureService` too, lets patch that as well.

File: app/models/reports/repository.rb (hunk 2 above)
Comment (rsanheim) [reply app/models/reports/repository.rb]: No need now, I patched ActiveSupport directly

**Review by kpethtel — APPROVED** [review]
It's not clear to me if anything should be done with Prabhanshu's comment about the `let` in the control rate service spec.

Also, if the rails team ever fixes this issue, will this fix become problematic? I assume not, since super is not being called when the force option is present. I suppose we'll have to keep an eye out for this fix in order to know when to remove this.

**Comment by rsanheim** [comment]
> Also, if the rails team ever fixes this issue, will this fix become problematic? I assume not, since super is not being called when the force option is present. I suppose we'll have to keep an eye out for this fix in order to know when to remove this.

Yup, this is the sort of thing we should be on top of when we do future Rails upgrades.  I'll leave a comment to that effect.

**Comment by rsanheim** [comment]
Did a deploy of this out to sandbox with the sandbox config change merged -- caching still works fine, and force_cache is also working correctly. Below are some example requests -- the ones with 0% cache hits were me hitting a report with `force_cache=1`, while the others were me hitting the report page normally after busting the cache.

```
  | DATE | HOST | SERVICE | CONTENT
-- | -- | -- | -- | --
  | Feb 23 18:09:31.371 | ec2-13-235-33-14.ap-south-1.compute.amazonaws.com | simple_server | cache hit rate: 100.0% stats: {:reads=>150, :hits=>150, :misses=>0, :missed_keys=>[]}
  | Feb 23 18:09:19.369 | ec2-13-235-33-14.ap-south-1.compute.amazonaws.com | simple_server | cache hit rate: 100.0% stats: {:reads=>150, :hits=>150, :misses=>0, :missed_keys=>[]}
  | Feb 23 18:09:08.366 | ec2-13-235-33-14.ap-south-1.compute.amazonaws.com | simple_server | cache hit rate: 0% stats: {:reads=>6, :hits=>0, :misses=>6, :missed_keys=>["ControlRateService/regions/state/55b3c607-11bf-4788-9cad-de875b5d5ca6/month/with_exclusions", "ControlRateService/regions/district/c18a088c-9cd7-46dd-aec1-f461f8cc658b/month/with_exclusions", "ControlRateService/regions/district/b1390cbe-aef9-452a-a295-7ad87884c777/month/with_exclusions", "ControlRateService/regions/district/856a3590-97b1-42af-badd-4464ce129caf/month/with_exclusions", "ControlRateService/regions/district/e1a3127d-5005-4cb3-b034-5dca0ed6e0cb/month/with_exclusions", "ControlRateService/regions/district/8f007315-1c74-4812-b213-50ab1a3e23b4/month/with_exclusions"]}
  | Feb 23 18:08:21.356 | ec2-13-235-33-14.ap-south-1.compute.amazonaws.com | simple_server | cache hit rate: 0% stats: {:reads=>6, :hits=>0, :misses=>6, :missed_keys=>["ControlRateService/regions/state/55b3c607-11bf-4788-9cad-de875b5d5ca6/month/with_exclusions", "ControlRateService/regions/district/c18a088c-9cd7-46dd-aec1-f461f8cc658b/month/with_exclusions", "ControlRateService/regions/district/b1390cbe-aef9-452a-a295-7ad87884c777/month/with_exclusions", "ControlRateService/regions/district/856a3590-97b1-42af-badd-4464ce129caf/month/with_exclusions", "ControlRateService/regions/district/e1a3127d-5005-4cb3-b034-5dca0ed6e0cb/month/with_exclusions", "ControlRateService/regions/district/8f007315-1c74-4812-b213-50ab1a3e23b4/month/with_exclusions"]}
  | Feb 23 18:07:36.346 | ec2-13-235-33-14.ap-south-1.compute.amazonaws.com | simple_server | cache hit rate: 100.0% stats: {:reads=>150, :hits=>150, :misses=>0, :missed_keys=>[]}
  | Feb 23 18:07:23.342 | ec2-13-235-33-14.ap-south-1.compute.amazonaws.com | simple_server | cache hit rate: 99.89% stats: {:reads=>928, :hits=>927, :misses=>1, :missed_keys=>["users/0e288fee-3de7-415a-8224-c158be403fd5/regions/index"]}
```

**Comment by rsanheim** [comment]
@prabhanshuguptagit this feels ready to merge, along w/ the config changes...just need to monitor Sentry and Datadog once it lands in prod.  But the sandbox verification gives me confidence.

**Pull request merged**